### PR TITLE
Adds normalization of localhost urls in targets page react

### PR DESCRIPTION
**Pull request by Harkishen-Singh**
Signed-off-by: Harkishen Singh <harkishensingh@hotmail.com>
Fixes: #6301 

targets_globalurl

### CODE REVIEW

**Comment by juliusv** [comment]
Hmm, interesting approach, thanks! My initial idea was to inject yet another global variable into the app (like we do in https://github.com/prometheus/prometheus/blob/master/web/ui/react-app/public/index.html#L12-L24), and then only do the URL transformations on the client side.

However, now that I see the global URL added as another field in the targets API, I wonder if that might not make sense too, since other consumers than the UI might want to externally probe the metrics endpoints. So it would have value beyond the UI. OTHO one could assume that API consumers know the external URL of the Prometheus server, so they could also do that kind of transformation themselves anyway. @brian-brazil what do you think?

**Comment by Harkishen-Singh** [comment]
> Hmm, interesting approach, thanks! My initial idea was to inject yet another global variable into the app (like we do in https://github.com/prometheus/prometheus/blob/master/web/ui/react-app/public/index.html#L12-L24), and then only do the URL transformations on the client side.
> 
> However, now that I see the global URL added as another field in the targets API, I wonder if that might not make sense too, since other consumers than the UI might want to externally probe the metrics endpoints. So it would have value beyond the UI. OTOH one could assume that API consumers know the external URL of the Prometheus server, so they could also do that kind of transformation themselves anyway. @brian-brazil what do you think?

@juliusv how about sending the globalURL in scrapeURL, if we do not have any problem showing it in the UI, a bit different from the old UI?

**Comment by juliusv** [comment]
> how about sending the globalURL in scrapeURL, if we do not have any problem showing it in the UI, a bit different from the old UI?

@Harkishen-Singh We are not allowed to do that because it would be a breaking change to the targets API. But even if we were allowed, I think it might be misleading to display a URL that is different from the one that Prometheus actually scrapes. So I think either your current way or the global injected variable way are ok. Just would like to get a second opinion on adding this new field to the API.

File: web/web.go [hunk 1]
```diff
@@ -289,6 +289,9 @@ func New(logger log.Logger, o *Options) *Handler {
 
 		ready: 0,
 	}
+	fmt.Println(o.ListenAddress)
```
Comment (juliusv) [inline web/web.go]: Remove debug statements

File: web/web.go [hunk 2]
```diff
@@ -297,6 +300,11 @@ func New(logger log.Logger, o *Options) *Handler {
 			return *h.config
 		},
 		o.Flags,
+		api_v1.Options{
```
Comment (juliusv) [inline web/web.go]: `Options` is a bit generic of a name in this context, given that all the other parameters to `NewAPI()` are also options of sorts. Maybe eventually it'd be good to put everything into one big options struct, like we do in `web.go` for the web handler. But how about for now, calling it `GlobalURLOptions` or something like that?

File: web/api/v1/api.go [hunk 3]
```diff
@@ -641,12 +683,22 @@ func (api *API) targets(r *http.Request) apiFuncResult {
...
-					Labels:             target.Labels().Map(),
-					ScrapePool:         key,
-					ScrapeURL:          target.URL().String(),
-					LastError:          lastErrStr,
+					DiscoveredLabels: target.DiscoveredLabels().Map(),
+					Labels:           target.Labels().Map(),
+					ScrapePool:       key,
+					ScrapeURL:        target.URL().String(),
+					GlobalURL:        globalURL.String(),
+					LastError: func() string {
```
Comment (juliusv) [inline web/api/v1/api.go]: This seems like an unrelated change that shouldn't be in this PR?

File: web/web.go [hunk 4]
```diff
@@ -297,6 +300,11 @@ func New(logger log.Logger, o *Options) *Handler {
...
+			Address: o.ListenAddress,
```
Comment (juliusv) [inline web/web.go]: Let's keep it called `ListenAddress` in the new struct as well.

File: web/api/v1/api.go [hunk 5]
```diff
@@ -175,6 +176,7 @@ type API struct {
 	config                func() config.Config
 	flagsMap              map[string]string
 	ready                 func(http.HandlerFunc) http.HandlerFunc
+	option                Options
```
Comment (juliusv) [inline web/api/v1/api.go]: `option` -> `options`, but see my other comment about naming it generally more specifically.

File: web/api/v1/api.go [hunk 6]
```diff
@@ -602,6 +607,43 @@ type TargetDiscovery struct {
 	DroppedTargets []*DroppedTarget `json:"droppedTargets"`
 }
 
+// Options contains fields for globalURl required in local targets.
```
Comment (juliusv) [inline web/api/v1/api.go]: I'd suggest:

```suggestion
// Options contains fields used for deriving the global URL for local targets.
```

File: web/web.go (hunk 2 above)
Comment (Harkishen-Singh) [reply web/web.go]: Sure. Done.

File: web/api/v1/api.go (hunk 5 above)
Comment (Harkishen-Singh) [reply web/api/v1/api.go]: we already have a func `options` that implements an interface. Hence, it would be conflicting. how about `opts`?

File: web/api/v1/api.go [hunk 7]
```diff
@@ -602,6 +607,43 @@ type TargetDiscovery struct {
...
+	Scheme        string
+}
+
+func getGlobalURL(u *url.URL, opts GlobalURLOptions) (*url.URL, error) {
+	host, port, err := net.SplitHostPort(u.Host)
+	if err != nil {
+		return u, err
+	}
+
+	for _, lhr := range []string{"127.0.0.1", "localhost"} {
```
[the remark this answers is not in the document]
Comment (juliusv) [reply web/api/v1/api.go]: FWIW this was adapted from https://github.com/prometheus/prometheus/blob/40dd13b07420a044cc1b0ca57f639c572583d9c1/web/web.go#L976, maybe it'd be good to share more code here, but I didn't press for that because we assume that the old UI (along with its code) is going away anyway.

[the remark this answers is not in the document]
Comment (juliusv) [reply web/api/v1/api.go]: (if we only end up keeping the new version of this code though, it would be good to keep the comments from the original)

File: web/api/v1/api.go (hunk 5 above)
Comment (juliusv) [reply web/api/v1/api.go]: Yeah, but `options` was too generic anyway, so now it'd be `globalURLOptions`?

File: web/ui/react-app/src/pages/targets/EndpointLink.tsx [hunk 8]
```diff
@@ -22,7 +23,7 @@ const EndpointLink: FC<EndpointLinkProps> = ({ endpoint }) => {
 
   return (
     <>
-      <a href={endpoint}>{`${protocol}//${host}${pathname}`}</a>
+      <a href={globalUrl}>{`${protocol}//${host}${pathname}`}</a>
```
[the remark this answers is not in the document]
Comment (juliusv) [reply web/ui/react-app/src/pages/targets/EndpointLink.tsx]: Nope, only the link target should be changed, as in the old UI, so this should be correct.

We want to display what the Prometheus server has configured, but if it actually leads to the Prometheus server itself (localhost), we want to un-break that link for the user.

File: web/api/v1/api.go (hunk 3 above)
Comment (Harkishen-Singh) [reply web/api/v1/api.go]: This is actually done to capture the errors from the `globalURL` as well. Without this, it might be a `go-lint` error since we are not catching the error reported by the `net` library.

File: web/api/v1/api.go (hunk 5 above)
Comment (Harkishen-Singh) [reply web/api/v1/api.go]: sure.

File: web/api/v1/api.go [hunk 9]
```diff
@@ -200,6 +205,7 @@ func NewAPI(
 	ar alertmanagerRetriever,
 	configFunc func() config.Config,
 	flagsMap map[string]string,
+	option GlobalURLOptions,
```
Comment (juliusv) [inline web/api/v1/api.go]: This `option` parameter still needs to be renamed to `globalURLOptions`.

File: web/api/v1/api.go [hunk 10]
```diff
@@ -602,6 +610,43 @@ type TargetDiscovery struct {
...
 }
 
+// GlobalURLOptions contains fields used for deriving the global URL for local targets.
+type GlobalURLOptions struct {
+	ListenAddress string
+	Host          string
+	Scheme        string
+}
+
+func getGlobalURL(u *url.URL, opts GlobalURLOptions) (*url.URL, error) {
```
Comment (juliusv) [inline web/api/v1/api.go]: As mentioned aside in https://github.com/prometheus/prometheus/pull/6794#discussion_r377932616, could you keep (re-add) the comments from the original implementation of this function?

Likely only this copy of the function will survive in the long run, so it would be good to have the comments here as well.

File: web/web_test.go [hunk 11]
```diff
@@ -391,8 +395,14 @@ func TestDebugHandler(t *testing.T) {
...
 		opts := &Options{
-			RoutePrefix: tc.prefix,
+			RoutePrefix:   tc.prefix,
+			ListenAddress: "somehost:9090",
+			ExternalURL: &url.URL{
+				Host:   "localhost.localdomain:9090",
+				Scheme: "http",
+			},
 		}
+		fmt.Println("opts is ", opts)
```
Comment (juliusv) [inline web/web_test.go]: Remove debug statement.

File: web/api/v1/api.go (hunk 3 above)
Comment (juliusv) [reply web/api/v1/api.go]: Ahhh, I see. The `LastError` field encodes scrape-related errors, so I don't think it's the right place for communicating errors related to computing the global URL (which is only relevant for UI purposes so far, and not scrape-related).

From what I can see, we pre-process both the external URL (in main.go) and the target's URL (in config.go) in such a way already that errors in `getGlobalURL()` are basically impossible (hopefully). So how about just failing the entire API call here in that case with an Internal Server Error, like in this example: https://github.com/prometheus/prometheus/blob/88af97366366fcbfff0af0f33d2da196a2afe629/web/api/v1/api.go#L1018-L1020

File: web/api/v1/api.go (hunk 9 above)
Comment (Harkishen-Singh) [reply web/api/v1/api.go]: Done

File: web/api/v1/api.go (hunk 10 above)
Comment (Harkishen-Singh) [reply web/api/v1/api.go]: Sure.

**Comment by juliusv** [comment]
:+1: Thanks, looks good!

I'll merge, but we'll also need to update the example response in the documentation in https://github.com/prometheus/prometheus/blob/master/docs/querying/api.md#targets - would you be up for a follow-up PR to do that?

**Comment by roidelapluie** [comment]
> +1 Thanks, looks good!
> 
> I'll merge, but we'll also need to update the example response in the documentation in https://github.com/prometheus/prometheus/blob/master/docs/querying/api.md#targets - would you be up for a follow-up PR to do that?

@juliusv @Harkishen-Singh Did that followup PR happen? We cut the RC in 3 days. Thanks.

**Comment by juliusv** [comment]
@roidelapluie I don't think it ever happened, I'll just do it.
